compiler/lib/deadcode_dgraph.ml

```diff
@@ -0,0 +1,461 @@
+        | _ -> instr)
+    | _ -> instr
+  in
+  let zero_block block =
+    (* Analyze block instructions *)
+    let body = List.map ~f:(fun (instr, loc) -> zero_instr instr, loc) block.body in
+    (* Analyze branch *)
+    let branch =
+      let last, loc = block.branch in
+      let last =
+        match last with
+        | Return x -> Return (zero_var x)
+        | Raise (_, _) | Stop -> last
+        | Branch cont -> Branch (zero_cont cont)
+        | Cond (x, cont1, cont2) -> Cond (x, zero_cont cont1, zero_cont cont2)
+        | Switch (x, a) ->
+            Switch (x, Array.map ~f:zero_cont a)
+        | Pushtrap (cont1, x, cont2, pcs) ->
+            Pushtrap (zero_cont cont1, x, zero_cont cont2, pcs)
+        | Poptrap cont -> Poptrap (zero_cont cont)
+      in
+      last, loc
+    in
+    { block with body; branch }
+  in
+  let blocks = prog.blocks |> Addr.Map.map zero_block in
+  { prog with blocks }
+
+module Print = struct
+  let live_to_string = function
+    | Live fields ->
+        "live { " ^ IntSet.fold (fun i s -> s ^ Format.sprintf "%d " i) fields "" ^ "}"
+    | Top -> "top"
+    | Dead -> "dead"
+
+  let print_uses uses =
+    Format.eprintf "Usages:\n";
+    Array.iteri
+      ~f:(fun i ds ->
+        Format.eprintf "v%d: { " i;
+        Var.Map.iter
+          (fun d k ->
+            Format.eprintf
+              "(%a, %s) "
+              Var.print
+              d
+              (match k with
+              | Compute -> "C"
+              | Propagate -> "P"))
+          ds;
+        Format.eprintf "}\n")
+      uses
+
+  let print_liveness live_vars =
+    Format.eprintf "Liveness:\n";
+    Array.iteri ~f:(fun i l -> Format.eprintf "v%d: %s\n" i (live_to_string l)) live_vars
+
+  let print_live_tbl live_table =
+    Format.eprintf "Liveness with dependencies:\n";
+    Var.Tbl.iter
+      (fun v l -> Format.eprintf "%a: %s\n" Var.print v (live_to_string l))
+      live_table
+end
+
+(** Add a sentinal variable declaration to the IR. The fresh variable is assigned to `undefined`. *)
+let add_sentinal p =
+  let sentinal = Var.fresh_n "sentinal" in
+  let undefined = Prim (Extern "%undefined", []) in
+  let instr, loc = Let (sentinal, undefined), Before 0 in
+  Code.prepend p [ instr, loc ], sentinal
+
+(** Run the liveness analysis and replace dead variables with the given sentinal. *)
+let f p sentinal global_info =
+  let t = Timer.make () in
+  let nv = Var.count () in
+  (* Compute definitions *)
+  let defs = definitions nv p in
+  (* Compute usages *)
+  let uses = usages nv p global_info in
+  (* Compute initial liveness *)
+  let pure_funs = Pure_fun.f p in
+  let live_vars = liveness nv p pure_funs global_info in
+  (* Propagate liveness to dependencies *)
+  let vars = variables uses in
+  let live_table = solver vars uses defs live_vars in
+  (* Zero out dead fields *)
+  let p = zero p sentinal live_table in
+  if debug ()
+  then (
+    Code.Print.program (fun _ _ -> "") p;
```

Comment:
You should move this before the call to `zero p sentinal live_table`, or use a different variable for the program before and after zeroing out dead fields.

Reply:
[5864030](https://github.com/ocsigen/js_of_ocaml/pull/1503/commits/58640303562f2f6680fdbaa0fe3f70c0ef5bd1e9)

Reply:
What I meant is that you are calling `Code.Print.program (fun _ _ -> "") p` twice with the same `p`.
I would expect one occurrence of this expression to be before `zero p sentinal live_table` and the other one after.

Reply:
Oh okay I misunderstood the first time. And I agree, it was like this earlier but I changed it at some point. Is [d44390f](https://github.com/ocsigen/js_of_ocaml/pull/1503/commits/d44390fe2292f06d49313425d689f3b4fae41f54) what you meant?